### PR TITLE
添加glm

### PR DESCRIPTION
**Pull request by FanGuojie**
接口和sklearn的一样，但如果有tol会给出提示
优化器有newton-cholesky，测试通过。lbfgs有近似实现，但测试下来发现误差比较大
总体结构：
一个基类GeneralizedLinearRegressor 实现的是线性回归，连接函数：IdentityLink 损失函数：HalfSquaredLoss
三个特殊分布的子类
PoissonRegressor，连接函数LogLink，损失函数：HalfPoissonLoss
GammaRegressor，连接函数LogLink，损失函数：HalfGammaLoss
TweedieRegressor，连接函数LogLink或IdentityLink（根据power值决定），损失函数：HalfTweedieLoss

仿真测试结果：（数据本地生成的，样本规模n_samples, n_features = 100, 5）
norm_diff 是明文和密态结果输出的回归参数二范数差

仿真结果：（数据模拟器生成的，原始数据分处不同机器节点上，用的是DATASET_MOCK_REGRESSION_BASIC）

小功能实现:多进程指定，热启动，sample_weight设置，L2正则化

### CODE REVIEW

**Comment by FanGuojie** [comment]
image

image

**Comment by tpppppub** [comment]
Hi，请不要将对 examples/data 里的修改 commit 到 PR 里来

File: examples/data/breast_cancer_a.csv
Comment (rivertalk) [inline examples/data/breast_cancer_a.csv:1]: 这个到底是什么diff呀，感觉这个文件不需要改变

Comment (rivertalk) [reply examples/data/breast_cancer_a.csv:1]: 可以试着用如下命令
```
git reset origin/master
git checkout -- examples
git add .
git commit -m "balabala"
```
单独把这个folder revert掉

File: sml/glm/glm_emul.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        """
+        X = jnp.concatenate((x1, x2), axis=1)
+        model = _GeneralizedLinearRegressor(solver="newton-cholesky")
+        model.fit(X, y)
+        return model.score(X, y), model.predict(X)
+
+    try:
+        # Specify the file paths for cluster and dataset
+        CLUSTER_ABY3_3PC = os.path.join('../../', emulation.CLUSTER_ABY3_3PC)
+        DATASET_MOCK_REGRESSION_BASIC = os.path.join('../../', emulation.DATASET_MOCK_REGRESSION_BASIC)
```
Comment (deadlywing) [inline sml/glm/glm_emul.py:58]: 需要更新一下spu代码哈，最近改动过emulator部分的api～

File: sml/glm/glm_emul.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+        # Create the emulator with specified mode and bandwidth/latency settings
+        emulator = emulation.Emulator(
+            CLUSTER_ABY3_3PC, mode, bandwidth=300, latency=20
+        )
+        emulator.up()
+
+        # Prepare the dataset using the emulator
+        (x1, x2), y = emulator.prepare_dataset(DATASET_MOCK_REGRESSION_BASIC)
```
Comment (deadlywing) [inline sml/glm/glm_emul.py:67]: 同理，请先自己读取数据，然后调用seal接口加载到PYU

File: sml/glm/glm.py
Comment (deadlywing) [inline sml/glm/glm.py:1]: spu整个项目都是使用bazel作为编排工具，麻烦补充一下bazel文件吧，方便统一测试～

可参考sml中其他文件夹下的BUILD～

Thanks

File: sml/glm/glm.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+# The TweedieRegressor class represents a generalized linear model with Tweedie distribution using JAX.
+class TweedieRegressor(_GeneralizedLinearRegressor):
+    def __init__(
+        self,
+        power=0.5,
+    ):
+        super().__init__()
+        # Ensure that the power is within the valid range for the Tweedie distribution
+        if power > 0:
```
Comment (deadlywing) [inline sml/glm/glm.py:174]: 这里对power做变换是为啥呀？

Comment (deadlywing) [inline sml/glm/glm.py:174]: 可以注明一下这里使用的是Extended Quasi-Likelihood，所以可以支持[1,2]以外的分布（likelihood funciton只在[1,2]上良定义）；

File: sml/glm/glm.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+        """
+        self.l2_reg_strength = alpha
+        self.fit_intercept = fit_intercept
+        self.solver = solver
+        self.max_iter = max_iter
+        self.warm_start = warm_start
+        self.verbose = verbose
+        if n_threads == 0:
+            os.environ["XLA_FLAGS"] = "--xla_cpu_multi_thread_eigen=true"
```
[the remark this answers is not in the document]
Comment (anakinxc) [reply sml/glm/glm.py:54]: > 这个是不是只对明文下运行有效呀？spu对这个环境变量也有特殊的处理么？ @anakinxc

没有，这个只对明文 cpu backend 有用

File: sml/glm/utils/solver.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+        -------
+        coef : array-like, shape (n_features,)
+            Solved coefficients.
+
+        """
+        X = super().solve(X, y)
+
+        # Use Cholesky factorization to solve linear systems
+        def cho_solve_wrapper(a, b):
+            return cho_solve(cho_factor(a), b[..., 0])
```
Comment (deadlywing) [inline sml/glm/utils/solver.py:114]: 为什么grad要先扩展axes然后再取第一个维度？不能直接传递grad么？

File: sml/glm/utils/loss.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            Predicted target data.
+
+        Returns:
+        -------
+        float
+            Average half squared loss value.
+
+        """
+        def loss_single_sample(y_t, y_p):
+            return jnp.mean((y_t - y_p) ** 2) / 2
```
Comment (deadlywing) [inline sml/glm/utils/loss.py:56]: 我理解y_t, y_p应该都是scalar?需要jnp.mean么？

File: examples/data/breast_cancer_a.csv
Comment (FanGuojie) [reply examples/data/breast_cancer_a.csv:1]: 好的，我是前天fork的，只是做了放置glm文件夹，不确定这个是怎么加进来的，这就删了，给您添麻烦了

File: sml/glm/glm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FanGuojie) [reply sml/glm/glm.py:54]: 我看jnp自动有调用多进程，所以这里还需要做这个多进程操作吗

File: sml/glm/glm_test.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        self.log_y = jnp.log(self.y)
+        self.round_logy = jnp.round(self.log_y)
+
+    def test_fit_predict(self):
+        """
+        Test the fitting and prediction functionality of different optimization algorithms for Generalized Linear Regression models.
+
+        """
+        # Set up the simulator
+        sim = spsim.Simulator.simple(3, spu_pb2.ProtocolKind.ABY3, spu_pb2.FieldType.FM64)
```
Comment (deadlywing) [inline sml/glm/glm_test.py:37]: FM64似乎不够，我本地测试需要FM128

File: sml/glm/glm.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        return y_pred
+
+    def score(self, X, y, sample_weight=None):
+        """
+        D^2 is the evaluation metric for the generalized linear regression model.
+        """
+
+        # Calculate the model's predictions
+        prediction = self.predict(X)
+        squared_error = lambda y_true, prediction: jnp.mean(
```
Comment (deadlywing) [inline sml/glm/glm.py:124]: deviation应该是一个和分布相关的计算，square loss应该只是针对normal distribution的deviation～

可以参考sklearn里的处理方式哈～

File: sml/glm/glm.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+        Initialize the generalized linear regression model.
+
+        Parameters:
+        ----------
+        fit_intercept : bool, optional
+            Whether to fit the intercept term, default is True.
+        alpha : float, optional
+            L2 regularization strength, default is 0 (no regularization).
+        solver : str, optional
+            Optimization algorithm, default is Newton-Cholesky. Supported values are "lbfgs" or "newton-cholesky".
```
Comment (deadlywing) [inline sml/glm/glm.py:33]: 关于优化器，glm明文有几个常见的算法：

1. lbfgs: 超线性收敛，明文下速度和精度都很好，极其依赖line search
2. newton / irls : 纯2阶；但前者理论上依赖line search，后者的优势在于根据分布可以给个较好的初始估计，对line search的依赖相对较弱 
3. 其余一阶算法：明文下使用较少（除非有L1需求），但理论上有收敛性保证

我理解如果要在纯密态下实现，irls或一阶算法可能会valid一些；其余可能需要reveal一些bits～

供参考

File: sml/glm/utils/solver.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+            self.gamma = rho_k_inv / (jnp.conj(y_k) @ y_k)
+            jnp.roll(self.s_history, -1, axis=0).at[-1, :].set(s_k)
+            jnp.roll(self.y_history, -1, axis=0).at[-1, :].set(y_k)
+            jnp.roll(self.rho_history, -1, axis=0).at[-1].set(rho_k)
+
+        return self.coef
+
+    def _two_loop_recursion(self, g_k):
+        his_size = len(self.rho_history)
+        curr_size = his_size
```
Comment (deadlywing) [inline sml/glm/utils/solver.py:219]: 似乎`curr_size`的作用和`his_size`一致？下面使用`curr_size`的地方也没体现出其语义，建议删除，只保留`his_size`

File: sml/glm/utils/solver.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+            a_his = a_his.at[i].set(a_i)
+            q = q - a_i * jnp.conj(self.y_history[i])
+
+        q = self.gamma * q
+
+        for j in range(curr_size):
+            i = his_size - curr_size + j
+            b_i = self.rho_history[i] * (self.y_history[i] @ q)
+            q = q + (a_his[i] - b_i) * self.s_history[i]
+        norm_q = jnp.linalg.norm(q)
```
Comment (deadlywing) [inline sml/glm/utils/solver.py:235]: 前面是标准的two-loop recursion没问题，最后这里的2-norm标准化是为什么呀？是参考的哪个package或者paper呢？

File: sml/glm/glm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/glm/glm.py:54]: hi，我建议不需要了，不然用户会以为密态下执行也能在这里控制并行度～

Thanks~

**Comment by FanGuojie** [comment]
更新unitest，以上问题都有解决

image

File: sml/glm/utils/solver.py (hunk 10 above)
Comment (FanGuojie) [reply sml/glm/utils/solver.py:219]: 之前是不一样的，但我过不了jit，也是一个动态改变的量，理论上它俩是不同的，因为都改成近似实现， 所以我暂时保留在这里

File: sml/glm/utils/solver.py (hunk 11 above)
Comment (FanGuojie) [reply sml/glm/utils/solver.py:235]: 最后是我手动调步长加的，理论上q我只需要知道它方向就够了吧，因为步长不能用line_search，所以我先在这里归一化，再后面调整步长的时候改的

File: sml/glm/utils/solver.py (hunk 10 above)
Comment (deadlywing) [reply sml/glm/utils/solver.py:219]: I see, 本来前maxcor的迭代次数应该是curr_size，现在相当于是为了避免动态性，所以都统一到maxcor次了，是这么理解嘛?

不过我乍看上去，应该也是正确的吧，因为初始化的都是0向量，所以你说的近似实现是近似在哪里呢？

File: sml/glm/utils/solver.py (hunk 11 above)
Comment (deadlywing) [reply sml/glm/utils/solver.py:235]: 那我觉得就没这个必要了，没有什么理论依据的话，反而给别人误解

File: sml/glm/glm_emul.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        # Specify the file paths for cluster and dataset
+        CLUSTER_ABY3_3PC = os.path.join('../../', emulation.CLUSTER_ABY3_3PC)
+        # Create the emulator with specified mode and bandwidth/latency settings
+        emulator = emulation.Emulator(
+            CLUSTER_ABY3_3PC, mode, bandwidth=300, latency=20
+        )
+        emulator.up()
+
+        # Run the proc_ncSolver function using both plaintext and encrypted data
+        raw_score, raw_result = proc_ncSolver(ppd.get(X), ppd.get(y))
```
Comment (deadlywing) [inline sml/glm/glm_emul.py:94]: 1. X,y本身就是明文的，不需要ppd.get
2. 我们推荐emul里直接使用emulator暴露的api，即emulator.seal(X, y)来把明文数据发送到PYU中

File: sml/glm/BUILD.bazel [hunk 13]
```diff
@@ -0,0 +1,35 @@
+load("@rules_python//python:defs.bzl", "py_binary", "py_library", "py_test")
+
+package(default_visibility = ["//visibility:public"])
+
+py_library(
+    name = "glm",
+    srcs = ["glm.py"],
+    deps = [
+        "//sml/utils:fxp_approx",
```
Comment (deadlywing) [inline sml/glm/BUILD.bazel:9]: 可以把实际不需要的deps去一下，包括下面的dataset_utils等

File: sml/glm/glm_emul.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
+import jax.numpy as jnp
+import numpy as np
+import sys
+import os
+
+# Add the parent directory to the system path to import custom modules
+sys.path.append('../../')
+import sml.utils.emulation as emulation
+import spu.utils.distributed as ppd
+from glm import _GeneralizedLinearRegressor, PoissonRegressor, GammaRegressor, TweedieRegressor
```
Comment (deadlywing) [inline sml/glm/glm_emul.py:10]: remove redundant import

Comment (deadlywing) [reply sml/glm/glm_emul.py:10]: 另外，还有两点：

1. 请clean 一下commit内容，把和glm无关的改动去掉
2. Python 代码建议使用 black+isort 进行格式化
